src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs

```diff
@@ -135,17 +141,70 @@ internal void RegisterForAdditionalFileChanges(EditCallback<AdditionalFileEdit>
         public void RegisterForSyntaxNotifications(SyntaxReceiverCreator receiverCreator)
         {
-            CheckIsEmpty(InfoBuilder.SyntaxReceiverCreator);
+            ValidateSyntaxReceiver(InfoBuilder);
             InfoBuilder.SyntaxReceiverCreator = receiverCreator;
         }
 
+        /// <summary>
+        /// Register a <see cref="SyntaxContextReceiverCreator"/> for this generator, which can be used to create an instance of an <see cref="ISyntaxContextReceiver"/>.
+        /// </summary>
+        /// <remarks>
+        /// This method allows generators to be 'syntax aware'. Before each generation the <paramref name="receiverCreator"/> will be invoked to create
+        /// an instance of <see cref="ISyntaxReceiver"/>. This receiver will have its <see cref="ISyntaxContextReceiver.OnVisitSyntaxNode(GeneratorSyntaxContext)"/> 
+        /// invoked for each syntax node in the compilation, allowing the receiver to build up information about the compilation before generation occurs.
+        /// 
+        /// During <see cref="ISourceGenerator.Execute(GeneratorExecutionContext)"/> the generator can obtain the <see cref="ISyntaxContextReceiver"/> instance that was
+        /// created by accessing the <see cref="GeneratorExecutionContext.SyntaxContextReceiver"/> property. Any information that was collected by the receiver can be
+        /// used to generate the final output.
+        /// 
+        /// A new instance of <see cref="ISyntaxContextReceiver"/> is created per-generation, meaning there is no need to manage the lifetime of the 
+        /// receiver or its contents.
+        /// </remarks>
+        /// <param name="receiverCreator">A <see cref="SyntaxContextReceiverCreator"/> that can be invoked to create an instance of <see cref="ISyntaxReceiver"/></param>
+        public void RegisterForSyntaxNotifications(SyntaxContextReceiverCreator receiverCreator)
+        {
+            ValidateSyntaxReceiver(InfoBuilder);
+            InfoBuilder.SyntaxReceiverWithContextCreator = receiverCreator;
+        }
+
         private static void CheckIsEmpty<T>(T x)
         {
             if (x is object)
             {
                 throw new InvalidOperationException(string.Format(CodeAnalysisResources.Single_type_per_generator_0, typeof(T).Name));
             }
         }
+
+        private static void ValidateSyntaxReceiver(GeneratorInfo.Builder builder)
+        {
+            if (builder.SyntaxReceiverCreator is object || builder.SyntaxReceiverWithContextCreator is object)
+            {
+                // only a single SyntaxReceiverCreator / SyntaxContextReceiverCreator can be registered per generator
+                throw new InvalidOperationException(string.Format(CodeAnalysisResources.Single_type_per_generator_0, $"{nameof(SyntaxReceiverCreator)} / {nameof(SyntaxContextReceiverCreator)}"));
```

Comment:
Why do we have this restriction? I can't exactly see a good reason to have both but it doesn't seem inherently wrong to do so either.

Reply:
We didn't specifically need it before, but now that we're using an adaptor you can't supply both.